reader.go

```diff
@@ -29,31 +66,770 @@ type Reader struct {
 	once  uint32
 	stctx context.Context
 	stats readerStats
 }
 
+// membership returns the group generationID and memberID of the reader.
+//
+// Only used when config.GroupID != ""
+func (r *Reader) membership() (generationID int32, memberID string) {
+	r.mutex.Lock()
+	generationID = r.generationID
+	memberID = r.memberID
+	r.mutex.Unlock()
+	return
+}
+
+// lookupCoordinator scans the brokers and looks up the address of the
+// coordinator for the group.
+//
+// Only used when config.GroupID != ""
+func (r *Reader) lookupCoordinator() (string, error) {
+	conn, err := r.broker()
+	if err != nil {
+		return "", fmt.Errorf("unable to connect to any broker for group, %v: %v\n", r.config.GroupID, err)
+	}
+	defer conn.Close()
+
+	out, err := conn.findCoordinator(findCoordinatorRequestV1{
+		CoordinatorKey: r.config.GroupID,
+	})
+	if err != nil {
+		return "", fmt.Errorf("unable to find coordinator for group, %v: %v", r.config.GroupID, err)
+	}
+
+	address := fmt.Sprintf("%v:%v", out.Coordinator.Host, out.Coordinator.Port)
+	return address, nil
+}
+
+// refreshCoordinator updates the value of r.address
+func (r *Reader) refreshCoordinator() (err error) {
+	const (
+		backoffDelayMin = 100 * time.Millisecond
+		backoffDelayMax = 1 * time.Second
+	)
+
+	for attempt := 0; true; attempt++ {
+		if attempt != 0 {
+			if !sleep(r.stctx, backoff(attempt, backoffDelayMin, backoffDelayMax)) {
+				return r.stctx.Err()
+			}
+		}
+
+		address, err := r.lookupCoordinator()
+		if err != nil {
+			continue
+		}
+
+		r.mutex.Lock()
+		oldAddress := r.address
+		r.address = address
+		r.mutex.Unlock()
+
+		if address != oldAddress {
+			r.withLogger(func(l *log.Logger) {
+				l.Printf("coordinator for group, %v, set to %v\n", r.config.GroupID, address)
+			})
+		}
+
+		break
+	}
+
+	return nil
+}
+
+// makeJoinGroupRequestV2 handles the logic of constructing a joinGroup
+// request
+func (r *Reader) makeJoinGroupRequestV2() (joinGroupRequestV2, error) {
+	_, memberID := r.membership()
+
+	request := joinGroupRequestV2{
+		GroupID:          r.config.GroupID,
+		MemberID:         memberID,
+		SessionTimeout:   int32(r.config.SessionTimeout / time.Millisecond),
+		RebalanceTimeout: int32(r.config.RebalanceTimeout / time.Millisecond),
+		ProtocolType:     defaultProtocolType,
+	}
+
+	for _, strategy := range allStrategies {
+		meta, err := strategy.ProtocolMetadata([]string{r.config.Topic})
+		if err != nil {
+			return joinGroupRequestV2{}, fmt.Errorf("unable to construct protocol metadata for member, %v: %v\n", strategy.ProtocolName(), err)
+		}
+
+		request.GroupProtocols = append(request.GroupProtocols, joinGroupRequestGroupProtocolV2{
+			ProtocolName:     strategy.ProtocolName(),
+			ProtocolMetadata: meta.bytes(),
+		})
+	}
+
+	return request, nil
+}
+
+// newMemberProtocolMetadata maps encoded member metadata ([]byte) into memberGroupMetadata
+func (r *Reader) newMemberProtocolMetadata(in []joinGroupResponseMemberV2) ([]memberGroupMetadata, error) {
```

Comment:
Should we rename this “make” as well? It returns a value from a make call, seems like it would make sense.